### PR TITLE
Python binding for quarters

### LINKED ISSUE #8676: [FEA] Series datetime quarter
For feature engineering tasks, I'd like to be able to determine what quarter of the year a date falls into, like I can in pandas. Pandas [implements](https://github.com/pandas-dev/pandas/blob/56657dccd313074209b11f03cc394cf021c69802/pandas/_libs/tslibs/timestamps.pyx#L661) this by extracting the month and doing a few binary ops on it.

```python
import pandas as pd
​
s = pd.Series(["2021-02-27", "2021-04-01"], dtype="datetime64[ms]")
s.dt.quarter
0    1
1    2
dtype: int64
```

This is also a Spark [function](https://spark.apache.org/docs/latest/api/sql/index.html#quarter), so we may want to implement this in libcudf.

### PR DESCRIPTION
**Pull request by shaneding**
Closes #8676.
This PR adds python bindings for #8779.

### CODE REVIEW

File: python/cudf/cudf/tests/test_datetime.py [hunk 1]
```diff
@@ -1297,3 +1297,34 @@ def test_is_leap_year():
...
+        "2000-12-31 04:00:00",
+        "1900-02-28 07:00:00",
+        "1800-03-14 07:30:00",
+        "2100-03-14 07:30:00",
+        "1970-01-01 00:00:00",
+        "1969-12-31 12:59:00",
+    ]
+
+    # Series
+    ps = pd.Series(data, dtype="datetime64[s]")
```
Comment (isVoid) [inline python/cudf/cudf/tests/test_datetime.py:1315]: We could parameterize the base unit too, I think.

**Comment by shaneding** [comment]
rerun tests

File: python/cudf/cudf/core/series.py [hunk 2]
```diff
@@ -6447,13 +6447,33 @@ def is_leap_year(self):
...
+        return Series._from_data(
+            ColumnAccessor({None: res}),
+            index=self.series._index,
+            name=self.series.name,
+        )
+
+    @property
+    def is_month_start(self):
+        """
+
```
Comment (isVoid) [inline python/cudf/cudf/core/series.py:6476]:
```suggestion
```

File: python/cudf/cudf/core/series.py [hunk 3]
```diff
@@ -6447,13 +6447,33 @@ def is_leap_year(self):
...
-    def is_month_start(self):
+    def quarter(self):
         """
-        Boolean indicator if the date is the first day of the month.
+        Integer indicator for which quarter of the year the date belongs in.
+
+        There are 4 quarters in a year. With the first quarter being from
+        January - March, second quarter being April - June, third quarter
+        being July - September and fourth quarter being October - December.
+                Boolean indicator if the date is the first day of the month.
```
Comment (isVoid) [inline python/cudf/cudf/core/series.py:6457]:
```suggestion
```

File: python/cudf/cudf/_lib/datetime.pyx [hunk 4]
```diff
@@ -67,3 +67,13 @@ def is_leap_year(Column col):
         c_result = move(libcudf_datetime.is_leap_year(col_view))
 
     return Column.from_unique_ptr(move(c_result))
+
+
+def extract_quarter(Column col):
```
Comment (galipremsagar) [inline python/cudf/cudf/_lib/datetime.pyx:72]: Can we add a short docstring describing what this cython binding does?

File: python/cudf/cudf/core/index.py [hunk 5]
```diff
@@ -2357,6 +2357,23 @@ def is_leap_year(self):
...
+        """
+        Integer indicator for which quarter of the year the date belongs in.
+
+        There are 4 quarters in a year. With the first quarter being from
+        January - March, second quarter being April - June, third quarter
+        being July - September and fourth quarter being October - December.
+
+        Returns
+        -------
+        Series
```
Comment (galipremsagar) [inline python/cudf/cudf/core/index.py:2371]:
```suggestion
        Int64Index
```

Could you also add some examples?

File: python/cudf/cudf/core/index.py [hunk 6]
```diff
@@ -2357,6 +2357,23 @@ def is_leap_year(self):
...
+        Integer indicating which quarter the date belongs to.
+        """
+        res = extract_quarter(self._values).fillna(False)
```
Comment (vyasr) [inline python/cudf/cudf/core/index.py:2374]: An integer value of 0 indicates Q0, i.e. Jan-Mar, correct? Doesn't filling nulls with `False` result in the equivalent of integer 0, which means that null values would get binned into Q0?

Comment (shaneding) [reply python/cudf/cudf/core/index.py:2374]: There is no such thing as Q0, it starts at Q1 so I don't think this would be an issue.

Comment (beckernick) [reply python/cudf/cudf/core/index.py:2374]: What happens in this scenario?

```python
import pandas as pd
​
s = pd.Series(["2021-01-01", None], dtype="datetime64[ms]")
s.dt.quarter
0    1.0
1    NaN
dtype: float64
```

Comment (shaneding) [reply python/cudf/cudf/core/index.py:2374]: It originally returned 0, but after removing the `.fillna(False)`, it replaced the 0 with `cudf.NA`, which I think is the correct behaviour?

Comment (beckernick) [reply python/cudf/cudf/core/index.py]: That LGTM

File: python/cudf/cudf/core/index.py [hunk 7]
```diff
@@ -2357,6 +2357,32 @@ def is_leap_year(self):
...
+        Integer indicating which quarter the date belongs to.
+
+        Examples
+        --------
+        >>> import pandas as pd
+        >>> import cudf
+        >>> pIndex = pd.DatetimeIndex(["2020-05-31 08:00:00",
+        ...    "1999-12-31 18:40:00"])
+        >>> cudf.from_pandas(pIndex).quarter
+        Int64Index([2, 4], dtype='int64')
```
Comment (isVoid) [inline python/cudf/cudf/core/index.py:2381]: Just one more nitpick: while we can, let's try to limit the example construction to using only cudf. This makes the example cleaner too.

```suggestion
        >>> import cudf
        >>> gIndex = cudf.DatetimeIndex(["2020-05-31 08:00:00",
        ...    "1999-12-31 18:40:00"])
        >>> gIndex.quarter
        Int64Index([2, 4], dtype='int64')
```

File: python/cudf/cudf/core/index.py (hunk 6 above)
Comment (vyasr) [reply python/cudf/cudf/core/index.py]: Yeah that was more my point, 0 shouldn't be the default output whether quarters are 0- or 1-indexed.

**Comment by shaneding** [comment]
rerun tests

**Comment by shaneding** [comment]
rerun tests

**Comment by shaneding** [comment]
rerun tests

**Review by isVoid — APPROVED** [review]
lgtm, good work!

File: python/cudf/cudf/core/index.py [hunk 8]
```diff
@@ -2283,6 +2283,31 @@ def is_leap_year(self):
...
+        Examples
+        --------
+        >>> import cudf
+        >>> gIndex = cudf.DatetimeIndex(["2020-05-31 08:00:00",
+        ...    "1999-12-31 18:40:00"])
+        >>> gIndex.quarter
+        Int64Index([2, 4], dtype='int64')
+        """
+        res = extract_quarter(self._values)
+        return Int64Index(res, dtype="int64")
```
Comment (galipremsagar) [inline python/cudf/cudf/core/index.py:2309]: Is it necessary to return an `int64`? The reason I'm asking is [1, 2, 3, 4] would easily fit in `int8` no?

File: python/cudf/cudf/core/series.py [hunk 9]
```diff
@@ -5944,13 +5944,39 @@ def is_leap_year(self):
...
+
+        Examples
+        -------
+        >>> import cudf
+        >>> s = cudf.Series(["2020-05-31 08:00:00","1999-12-31 18:40:00"],
+        ...     dtype="datetime64[ms]")
+        >>> s.dt.quarter
+        0    2
+        1    4
+        dtype: int16
```
Comment (galipremsagar) [inline python/cudf/cudf/core/series.py:5968]: Here aswell.

File: python/cudf/cudf/tests/test_datetime.py [hunk 10]
```diff
@@ -1300,6 +1300,38 @@ def test_is_leap_year():
...
+    dtype = "datetime64[s]"
+
+    # Series
+    ps = pd.Series(data, dtype=dtype)
+    gs = cudf.from_pandas(ps)
+
+    expect = ps.dt.quarter
+    got = gs.dt.quarter.astype(np.int64)
+
+    assert_eq(expect, got)
```
Comment (galipremsagar) [inline python/cudf/cudf/tests/test_datetime.py:1323]:
```suggestion
    got = gs.dt.quarter

    assert_eq(expect, got, check_dtype=False)
```

We could use `check_dtype=False` if we don't have to match pandas here.

**Review by galipremsagar — CHANGES_REQUESTED** [review]
I think we could probably save some memory that would be going to waste here. 

@beckernick Is it a necessity, in this case, we should be matching pandas return dtype(i.e., `int64`)? 

The related libcudf method also seems to be returning `int16`, instead of `int8`: https://github.com/rapidsai/cudf/pull/8779/files#diff-76e529bce1f51fca7ea3193f09c6e95906410a82bcc03a9271553b8527468052R218

@shaneding any reason why we choose `int16` over the lowest possible `dtype`?

File: python/cudf/cudf/core/index.py (hunk 8 above)
Comment (shaneding) [reply python/cudf/cudf/core/index.py:2309]: yes it would, but I think this was done to match Pandas behaviour.

Comment (shaneding) [reply python/cudf/cudf/core/index.py:2309]: Would it be preferable here to use an `Int8Index` here instead?

File: python/cudf/cudf/tests/test_datetime.py (hunk 10 above)
Comment (shaneding) [reply python/cudf/cudf/tests/test_datetime.py:1323]: @beckernick would it be preferable here to mimic pandas completely, or to use something more memory efficient?

Comment (beckernick) [reply python/cudf/cudf/tests/test_datetime.py:1323]: We should save memory when we have guarantees around the maximum size, as we're often more memory constrained. Thanks for flagging this @galipremsagar @shaneding

Comment (galipremsagar) [reply python/cudf/cudf/tests/test_datetime.py:1323]: Sure. @shaneding, in that case can we leave libcudf API untouched and return `int8` types at python level.

File: python/cudf/cudf/core/index.py (hunk 8 above)
Comment (galipremsagar) [reply python/cudf/cudf/core/index.py:2309]: Yes, lets return `Int8Index` and ignore dtype checks in pytests

File: python/cudf/cudf/tests/test_datetime.py (hunk 10 above)
Comment (beckernick) [reply python/cudf/cudf/tests/test_datetime.py:1323]: Shane shared that discussion with me as well. In this case, I think the performance penalty of typecasting for the `quarter` function is worth saving memory in Python workloads. 

We may want to think holistically about this discussion on Python side, if there hasn't already been a general pattern defined

Comment (galipremsagar) [reply python/cudf/cudf/tests/test_datetime.py:1323]: > We may want to think holistically about this discussion on Python side, if there hasn't already been a general pattern defined

We do this kind of memory saving on the python side for some normalization ops, replace ops, etc. with the help of following APIs:
`cudf.utils.dtypes.min_column_type`: https://github.com/rapidsai/cudf/blob/branch-21.10/python/cudf/cudf/utils/dtypes.py#L374
`cudf.utils.dtypes.min_float_dtype`: https://github.com/rapidsai/cudf/blob/branch-21.10/python/cudf/cudf/utils/dtypes.py#L404

We do this in some APIs already for `replace`, and all the APIs that use `_apply_support_method_axis_1` methods.

But in this specific case we could just hard-code `int8` as dtype.

File: python/cudf/cudf/core/index.py [hunk 11]
```diff
@@ -2283,6 +2283,31 @@ def is_leap_year(self):
...
+        """
+        Integer indicator for which quarter of the year the date belongs in.
+
+        There are 4 quarters in a year. With the first quarter being from
+        January - March, second quarter being April - June, third quarter
+        being July - September and fourth quarter being October - December.
+
+        Returns
+        -------
+        Int64Index
```
Comment (galipremsagar) [inline python/cudf/cudf/core/index.py:2297]:
```suggestion
        Int8Index
```

File: python/cudf/cudf/core/series.py [hunk 12]
```diff
@@ -5944,13 +5944,39 @@ def is_leap_year(self):
...
+        0    2
+        1    4
+        dtype: int8
+        """
+        res = libcudf.datetime.extract_quarter(self.series._column)
+        return Series._from_data(
+            ColumnAccessor({None: res}),
+            index=self.series._index,
+            name=self.series.name,
+        ).astype(np.int8)
```
Comment (galipremsagar) [inline python/cudf/cudf/core/series.py:5975]:
```suggestion
        res = libcudf.datetime.extract_quarter(self.series._column).astype(np.int8)
        return Series._from_data(
            ColumnAccessor({None: res}),
            index=self.series._index,
            name=self.series.name,
        )
```

File: python/cudf/cudf/tests/test_datetime.py [hunk 13]
```diff
@@ -1300,6 +1300,38 @@ def test_is_leap_year():
...
+    assert_eq(expect, got, check_dtype=False)
+
+    # DatetimeIndex
+    pIndex = pd.DatetimeIndex(data)
+    gIndex = cudf.from_pandas(pIndex)
+
+    expect2 = pIndex.quarter
+    got2 = gIndex.quarter
+
+    assert_eq(expect2.values, got2.values, check_dtype=False)
```
Comment (galipremsagar) [inline python/cudf/cudf/tests/test_datetime.py:1332]:
```suggestion
    assert isinstance(got2, cudf.Int8Index)
    assert_eq(expect2.values, got2.values, check_dtype=False)
```

**Review by galipremsagar — CHANGES_REQUESTED** [review]

File: python/cudf/cudf/core/series.py (hunk 12 above)
Comment (vyasr) [reply python/cudf/cudf/core/series.py:5975]: Small note, explicit ColumnAccessor construction is no longer necessary. Passing a dict `{None: res}` is fine.

Comment (galipremsagar) [reply python/cudf/cudf/core/series.py:5975]: Thanks @vyasr, updated the code suggestion.

**Review by galipremsagar — APPROVED** [review]

**Comment by galipremsagar** [comment]
@gpucibot merge

**Pull request merged**